File: src/m/solve/solve.py
from datetime import datetime
import os

from ismodelselfconsistent import ismodelselfconsistent
from loadresultsfromcluster import loadresultsfromcluster
from marshall import marshall
from pairoptions import pairoptions
from preqmu import *
from waitonlock import waitonlock


def solve(md, solutionstring, *args):
    """solve - apply solution sequence for this model

    Usage:
        md = solve(md, solutionstring, varargin)

    where varargin is a list of paired arguments of string OR enums

    Solution types available comprise:
    - 'Stressbalance'        or 'sb'
    - 'Masstransport'        or 'mt'
    - 'Oceantransport'       or 'oceant'
    - 'Thermal'              or 'th'
    - 'Steadystate'          or 'ss'
    - 'Transient'            or 'tr'
    - 'Balancethickness'     or 'mc'
    - 'BalancethicknessSoft' or 'mcsoft'
    - 'Balancevelocity'      or 'bv'
    - 'BedSlope'             or 'bsl'
    - 'SurfaceSlope'         or 'ssl'
    - 'Hydrology'            or 'hy'
    - 'DamageEvolution'      or 'da'
    - 'Gia'                  or 'gia'
    - 'Love'                 or 'lv'
    - 'Esa'                  or 'esa'
    - 'Sampling'             or 'smp'

    Extra options:
    - loadonly         : do not solve, only load results
    - runtimename      : true or false (default is true); makes name unique
    - checkconsistency : 'yes' or 'no' (default is 'yes'); checks consistency 
                         of model
    - restart          : directory name (relative to the execution directory) 
                         where the restart file is located

    Examples:
        md = solve(md, 'Stressbalance')
        md = solve(md, 'sb')
    """

    # Recover and process solve options
    if solutionstring.lower() == 'sb' or solutionstring.lower() == 'stressbalance':
        solutionstring = 'StressbalanceSolution'
    elif solutionstring.lower() == 'mt' or solutionstring.lower() == 'masstransport':
        solutionstring = 'MasstransportSolution'
    elif solutionstring.lower() == 'oceant' or solutionstring.lower() == 'oceantransport':
        solutionstring = 'OceantransportSolution'
    elif solutionstring.lower() == 'th' or solutionstring.lower() == 'thermal':
        solutionstring = 'ThermalSolution'
    elif solutionstring.lower() == 'st' or solutionstring.lower() == 'steadystate':
        solutionstring = 'SteadystateSolution'
    elif solutionstring.lower() == 'tr' or solutionstring.lower() == 'transient':
        solutionstring = 'TransientSolution'
    elif solutionstring.lower() == 'mc' or solutionstring.lower() == 'balancethickness':
        solutionstring = 'BalancethicknessSolution'
    elif solutionstring.lower() == 'bv' or solutionstring.lower() == 'balancevelocity':
        solutionstring = 'BalancevelocitySolution'
    elif solutionstring.lower() == 'bsl' or solutionstring.lower() == 'bedslope':
        solutionstring = 'BedSlopeSolution'
    elif solutionstring.lower() == 'ssl' or solutionstring.lower() == 'surfaceslope':
        solutionstring = 'SurfaceSlopeSolution'
    elif solutionstring.lower() == 'hy' or solutionstring.lower() == 'hydrology':
        solutionstring = 'HydrologySolution'
    elif solutionstring.lower() == 'da' or solutionstring.lower() == 'damageevolution':
        solutionstring = 'DamageEvolutionSolution'
    elif solutionstring.lower() == 'gia' or solutionstring.lower() == 'gia':
        solutionstring = 'GiaSolution'
    elif solutionstring.lower() == 'lv' or solutionstring.lower() == 'love':
        solutionstring = 'LoveSolution'
    elif solutionstring.lower() == 'esa':
        solutionstring = 'EsaSolution'
    elif solutionstring.lower() == 'smp' or solutionstring.lower() == 'sampling':
        solutionstring = 'SamplingSolution'
    else:
        raise ValueError('solutionstring {} not supported!'.format(solutionstring))
    options = pairoptions('solutionstring', solutionstring, *args)

    # Recover some fields
    md.private.solution = solutionstring
    cluster = md.cluster
    if options.getfieldvalue('batch', 'no') == 'yes':
        batch = 1
    else:
        batch = 0

    # Check model consistency
    if options.getfieldvalue('checkconsistency', 'yes') == 'yes':
        if md.verbose.solution:
            print('checking model consistency')
        ismodelselfconsistent(md)

    # If we are restarting, actually use the provided runtime name
    restart = options.getfieldvalue('restart', '')
    # First, build a runtime name that is unique
    if restart == 1:
        pass # Leave the runtimename as is
    else:
        if not isempty(restart):
            md.private.runtimename = restart
        else:
            if options.getfieldvalue('runtimename', True):
                c = datetime.now()
                md.private.runtimename = '%s-%02i-%02i-%04i-%02i-%02i-%02i-%i' % (md.miscellaneous.name, c.month, c.day, c.year, c.hour, c.minute, c.second, os.getpid())
            else:
                md.private.runtimename = md.miscellaneous.name

    # If running QMU analysis, some preprocessing of Dakota files using model 
    # fields needs to be carried out
    if md.qmu.isdakota:
        md = preqmu(md, options)

    # Do we load results only?
    if options.getfieldvalue('loadonly', False):
        md = loadresultsfromcluster(md)
        return md

    # Write all input files
    marshall(md) # bin file
    md.toolkits.ToolkitsFile(md.miscellaneous.name + '.toolkits') # toolkits file
    cluster.BuildQueueScript(md.private.runtimename, md.miscellaneous.name, md.private.solution, md.settings.io_gather, md.debug.valgrind, md.debug.gprof, md.qmu.isdakota, md.transient.isoceancoupling) # queue file

    # Upload all required files
    modelname = md.miscellaneous.name
    filelist = [modelname + '.bin', modelname + '.toolkits']

    if ispc():
        filelist.append(modelname + '.bat')
    else:
        filelist.append(modelname + '.queue')

    if md.qmu.isdakota:
        filelist.append(modelname + '.qmu.in')

    if isempty(restart):
        print('uploading input files')
        cluster.UploadQueueJob(md.miscellaneous.name, md.private.runtimename, filelist)

    # Launch job
    print('launching solution sequence')
    cluster.LaunchQueueJob(md.miscellaneous.name, md.private.runtimename, filelist, restart, batch)

    # Return if batch
    if batch:
        if md.verbose.solution:
            print('batch mode requested: not launching job interactively')
            print('launch solution sequence on remote cluster by hand')
        return md

    # Wait on lock
    if md.settings.waitonlock > 0:
        # Wait for done file
        done = waitonlock(md)
        if md.verbose.solution:
            print('loading results from cluster')
        md = loadresultsfromcluster(md)
    elif md.settings.waitonlock == 0:
        print('Model results must be loaded manually with md = loadresultsfromcluster(md).')

    return md

File: src/m/solve/solve.m
function md=solve(md,solutionstring,varargin)
%SOLVE - apply solution sequence for this model
%
%   Usage:
%      md=solve(md,solutionstring,varargin)
%
%   where varargin is a lit of paired arguments of string OR enums
%
%   Solution types available comprise:
%   - 'Stressbalance'        or 'sb'
%   - 'Masstransport'        or 'mt'
%   - 'Oceantransport'       or 'oceant'
%   - 'Thermal'              or 'th'
%   - 'Steadystate'          or 'ss'
%   - 'Transient'            or 'tr'
%   - 'Balancethickness'     or 'mc'
%   - 'Balancethickness2'
%   - 'BalancethicknessSoft' or 'mcsoft'
%   - 'Balancevelocity'      or 'bv'
%   - 'BedSlope'             or 'bsl'
%   - 'SurfaceSlope'         or 'ssl'
%   - 'Hydrology'            or 'hy'
%   - 'DamageEvolution'      or 'da'
%   - 'Gia'                  or 'gia'
%   - 'Love'                 or 'lv'
%   - 'Esa'                  or 'esa'
%   - 'Sampling'             or 'smp'
%
%   Extra options:
%   - loadonly         : do not solve, only load results
%   - runtimename      : true or false (default is true); makes name unique
%   - batch            : create input files but do not submit job
%   - checkconsistency : 'yes' or 'no' (default is 'yes'); checks consistency of model
%   - restart          : directory name (relative to the execution directory) 
%                        where the restart file is located
%
%   Examples:
%      md=solve(md,'Stressbalance');
%      md=solve(md,'sb');

if ~ischar(solutionstring)
	error('ISSM''s solve function only accepts strings for solution sequences. Type help solve to get a list of supported solutions');
end

%recover and process solve options
if strcmpi(solutionstring,'sb') || strcmpi(solutionstring,'Stressbalance')
	solutionstring = 'StressbalanceSolution';
elseif strcmpi(solutionstring,'mt') || strcmpi(solutionstring,'Masstransport')
	solutionstring = 'MasstransportSolution';
elseif strcmpi(solutionstring,'oceant') || strcmpi(solutionstring,'Oceantransport')
	solutionstring = 'OceantransportSolution';
elseif strcmpi(solutionstring,'th') || strcmpi(solutionstring,'Thermal')
	solutionstring = 'ThermalSolution';
elseif strcmpi(solutionstring,'ss') || strcmpi(solutionstring,'Steadystate') 
	solutionstring = 'SteadystateSolution';
elseif strcmpi(solutionstring,'tr') || strcmpi(solutionstring,'Transient')
	solutionstring = 'TransientSolution';
elseif strcmpi(solutionstring,'mc') || strcmpi(solutionstring,'Balancethickness')
	solutionstring = 'BalancethicknessSolution';
elseif strcmpi(solutionstring,'Balancethickness2')
	solutionstring = 'Balancethickness2Solution';
elseif strcmpi(solutionstring,'mcsoft') || strcmpi(solutionstring,'BalancethicknessSoft')
	solutionstring = 'BalancethicknessSoftSolution';
elseif strcmpi(solutionstring,'bv') || strcmpi(solutionstring,'Balancevelocity')
	solutionstring = 'BalancevelocitySolution';
elseif strcmpi(solutionstring,'bsl') || strcmpi(solutionstring,'BedSlope')
	solutionstring = 'BedSlopeSolution';
elseif strcmpi(solutionstring,'ssl') || strcmpi(solutionstring,'SurfaceSlope')
	solutionstring = 'SurfaceSlopeSolution';
elseif strcmpi(solutionstring,'hy') || strcmpi(solutionstring,'Hydrology')
	solutionstring = 'HydrologySolution';
elseif strcmpi(solutionstring,'da') || strcmpi(solutionstring,'DamageEvolution')
	solutionstring = 'DamageEvolutionSolution';
elseif strcmpi(solutionstring,'gia') || strcmpi(solutionstring,'Gia')
	solutionstring = 'GiaSolution';
elseif strcmpi(solutionstring,'lv') || strcmpi(solutionstring,'Love')
	solutionstring = 'LoveSolution';
elseif strcmpi(solutionstring,'esa') || strcmpi(solutionstring,'Esa')
	solutionstring = 'EsaSolution';
elseif strcmpi(solutionstring,'smp') || strcmpi(solutionstring,'Sampling')
	solutionstring = 'SamplingSolution';    
else
	error(['solutionstring ' solutionstring ' not supported!']);
end
options=pairoptions(varargin{:},'solutionstring',solutionstring);

%If we are restarting, actually use the provided runtime name:
restart=getfieldvalue(options,'restart','');
if restart==1 
	%Leave the runtimename as is
else
	if ~isempty(restart)
		md.private.runtimename=restart;
	elseif getfieldvalue(options,'runtimename',true)
		c=clock;
		md.private.runtimename=sprintf('%s-%02i-%02i-%04i-%02i-%02i-%02i-%i',md.miscellaneous.name,c(2),c(3),c(1),c(4),c(5),floor(c(6)),feature('GetPid'));
	else
		md.private.runtimename=md.miscellaneous.name;
	end
end

%Do we load results only?
if getfieldvalue(options,'loadonly',false)
	md=loadresultsfromcluster(md);
	return;
end

%recover some fields
cluster=md.cluster;
if strcmpi(getfieldvalue(options,'batch','no'),'yes')
	batch=1;
else
	batch=0;
end

%check model consistency
if strcmpi(getfieldvalue(options,'checkconsistency','yes'),'yes')
	md.private.solution=solutionstring;
	if md.verbose.solution
		disp('checking model consistency');
	end
	ismodelselfconsistent(md);
end

%Prepare directory in execution
if strcmpi(cluster.name, oshostname())
	localexecdir = cluster.executionpath;
else
	localexecdir = [issmdir() '/execution/'];
end
if ~exist(localexecdir, 'dir')
	error(['Could not find directory ' issmdir() '/execution/']);
elseif numel(dir(localexecdir))>200
	warning([localexecdir ' has more than 200 subdirectories. Consider cleaning up your execution directory'])
end
root = [localexecdir '/' md.private.runtimename];
if exist(root, 'dir')
	rmdir(root, 's');
end
mkdir(root);

%Figure out executable
executable = 'issm.exe';
if md.qmu.isdakota
	md=preqmu(md,options);
	movefile([md.miscellaneous.name '.qmu.in'], [root '/' md.miscellaneous.name '.qmu.in']);
	dakota_version_str = IssmConfig('_DAKOTA_VERSION_'); dakota_ver = str2num(dakota_version_str(1:3));
	if(dakota_ver >= 6), executable = 'issm_dakota.exe'; end
end

%Write all input files
basename = [root '/' md.miscellaneous.name];
marshall(md, [basename '.bin']);                               % bin file
ToolkitsFile(md.toolkits, [basename '.toolkits']);             % toolkits file
BuildQueueScript(cluster, md, [basename '.queue'], executable);% queue file

%List all required files
filelist  = {[basename '.bin'] [basename '.toolkits']};
if ispc
	filelist{end+1} = [basename '.bat'];
else
	filelist{end+1} = [basename '.queue'];
end
if md.qmu.isdakota
	filelist{end+1} = [basename '.qmu.in'];
end
if isprop(cluster, 'interactive') && cluster.interactive
	fid=fopen([basename '.errlog'],'w'); fclose(fid);
	fid=fopen([basename '.outlog'],'w'); fclose(fid);
	filelist{end+1} = [basename '.outlog'];
	filelist{end+1} = [basename '.errlog'];
end

%Upload input files if necessary
if isempty(restart) && ~strcmpi(cluster.name, oshostname())
   disp('uploading input files')
   UploadQueueJob(cluster, md.miscellaneous.name, md.private.runtimename, filelist);
end

%launch queue job: 
disp('launching solution sequence')
LaunchQueueJob(cluster,md.miscellaneous.name,md.private.runtimename,filelist,restart,batch);

%return if batch:
if batch
	disp('batch mode requested: not launching job interactively');
	disp('launch solution sequence on remote cluster by hand');
	return;
end

%wait on lock
if md.settings.waitonlock==0
	disp('Model results must be loaded manually with md=loadresultsfromcluster(md);');
	return
elseif isnan(md.settings.waitonlock)
	disp('solution launched on remote cluster. Log in to detect job completion.');
	choice=input('Is the job successfully completed? (y/n)','s');
	if ~strcmp(choice,'y')
		disp('Results not loaded... exiting'); 
		return;
	else
	end
elseif md.settings.waitonlock>0
	done = waitonlock(md);
else
	error('not supported');
end

if md.verbose.solution; disp('loading results from cluster'); end
md=loadresultsfromcluster(md);
